lasagne/layers/special.py

```diff
@@ -15,6 +16,57 @@
 ]
 
 
+class ExpressionLayer(Layer):
+    """
+    This layer provides boilerplate for a custom layer that applies a
+    simple transformation to the input.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+
+    function : callable
+        A function to be applied to the output of the previous layer.
+
+    output_shape : None, callable, tuple, or 'auto'
+        Specifies the output shape of this layer. If a tuple, this fixes the
+        output shape for any input shape (the tuple can contain None if some
+        dimensions may vary). If a callable, it should return the calculated
+        output shape given the input shape. If None, the output shape is
+        assumed to be the same as the input shape. If 'auto', an attempt will
+        be made to automatically infer the correct output shape.
```

Comment:
> When will output_shape='auto' fail?

for example, this will give a `NotImplementedError`:

    W = theano.shared(np.zeros((1, 1, 3, 3)).astype('float32'))
    def f(X):
        return T.nnet.conv2d(X, W)
    l2 = ExpressionLayer(l, f, output_shape='auto')